File: 122.买卖股票的最佳时机-ii.cpp
/*
 * @lc app=leetcode.cn id=122 lang=cpp
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start
class Solution {
public:
    // 贪心
    int maxProfit(vector<int>& prices) {
        int sum = 0;
        for(int i = 1; i < prices.size(); i++) {
            if(prices[i - 1] < prices[i]) {
                sum += prices[i] - prices[i - 1];
            }
        }
        return sum;
    }
};
// @lc code=end

File: 122.买卖股票的最佳时机-ii.go
/*
 * @lc app=leetcode.cn id=122 lang=golang
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start


func maxProfit(prices []int) int {
	size := len(prices)
	sum := 0
	has := 0
	h := false
	for i := 0; i < size-1; i++ {
		if prices[i] < prices[i+1] {
			// 今天比明天的要小
			if h {

			} else {
				has = prices[i]
				h = true
			}
		} else {
			// 明天降价
			if h {
				// 持有，先卖
				sum += prices[i] - has
				h = false
				has = 0
			}
		}
	}
	if h {
		sum += prices[size-1] - has
	}
	return sum
}

// @lc code=end

